BJS2-66474

Artpyak opened the pull request:
Задача: https://faang-school.atlassian.net/browse/BJS2-66474

Реализована заготовка для отправки уведомлений через telegram бота

Artpyak committed:
BJS2-66474
Реализована заготовка для отправки уведомлений через telegram бота

Graccio commented on src/main/java/faang/school/notificationservice/service/TelegramService.java:
```diff
@@ -0,0 +1,55 @@
+            telegramClient.execute(sendMessage);
+        } catch (TelegramApiException e) {
+            log.error("Error when sending a message to the user {}: {}", user.getId(), e.getMessage(), e);
+            throw new TelegramMessageException("Error when sending a message to the user", user.getId(), e);
```
Не уверен, но вроде бы из-за того, что ты ловишь `TelegramApiException` и вместо него выбрасываешь кастомное `TelegramMessageException` - у тебя не будут работать ретраи, так как в них ты привязываешься к `TelegramApiException`.

Artpyak committed:
BJS2-66474
Убрал кастомное исключение
Добавил аспект для работы ретрая

LauraLomberg commented on README.md:
```diff
@@ -107,3 +107,53 @@ RESTful приложения калькулятор с единственным
+
+## Важно!
+Текущий код класса `TelegramService` — это **заготовка**, работающая только для пользователей, уже написавших боту. 
+Для полной функциональности требуется реализовать механизм подписки.
```
Мне нравится твое предложение по решению проблемы. Считаю, что использовать userId, как chatId некорректно, хотя в условиях к задаче так и написано. Но они точно не будут совпадать и ничего не будет работать.

LauraLomberg commented on src/main/java/faang/school/notificationservice/service/TelegramService.java:
```diff
@@ -0,0 +1,54 @@
+
+    public TelegramService(@Value("${telegram.bot.token}") String token) {
+        this.telegramClient = new OkHttpTelegramClient(token);
+    }
```
Необязательно, но можно вынести TelegramClient в @Bean в отдельный конфигурационный класс. Это улучшит тестируемость и расширяемость.

Artpyak replied:
А это отличная идея) Так и сделаю)

Artpyak committed:
BJS2-66474
Вынес TelegramClient

Artpyak committed:
BJS2-66474
добавил более новую версию spring-retry 2.0.11

PavelEzhkov commented on README.md:
```diff
@@ -107,3 +107,53 @@ RESTful приложения калькулятор с единственным
 * Redis connectivity
 * ...
+
+# Telegram Notification Service
```
Это лучше чем я думал, очень хорошо описал

PavelEzhkov commented on src/main/java/faang/school/notificationservice/service/NotificationService.java:
```diff
@@ -1,10 +1,11 @@
 public interface NotificationService {
 
-    void send(UserDto user, String message);
+    void send(UserDto user, String message) throws TelegramApiException;
```
не надо менять этот контракт, у нас еще 2 сервиса, которые его имплементят

PavelEzhkov commented on src/main/java/faang/school/notificationservice/service/TelegramService.java:
```diff
@@ -0,0 +1,50 @@
+            telegramClient.execute(sendMessage);
+        } catch (TelegramApiException e) {
+            log.error("Error when sending a message to the user {}: {}", user.getId(), e.getMessage(), e);
+            throw e;
```
давай сделаем свою ошибку, где просто скажем что не смогли отправить, хотя конечно мне нравится твоя идея тем, что можно было бы сделать простую отправку в на почту или в смс если телега не отвечает

Artpyak committed:
BJS2-66474
1. Отменил изменения NotificationService.
2. Добавил кастомное исключение.
3. Изменил тест.
